tools/test-framework/src/types/topology.rs

```diff
@@ -0,0 +1,145 @@
+/*!
+   The Topology defines how chains are interconnected when more than two are used in tests.
+   This setup is managed by the [`crate::bootstrap::nary::chain::boostrap_chains_with_any_nodes`]
+   function.
+
+   Connections are established by examining the existing clients, and channels are created based
+   on these connections. Therefore, to define the topology of the chains, it is sufficient to
+   create the appropriate set of clients.
+
+   Example: Linear Topology
+
+   For a linear topology between chains A, B, and C, the clients are created as follows:
+   * Chain A: a client referencing chain B
+   * Chain B: a client referencing chain B and a client referencing chain C
+   * Chain C: a client referencing chain B
+
+   This setup ensures that:
+
+   * Chain A is connected to Chain B.
+   * Chain B is connected to both Chain A and Chain C.
+   * Chain C is connected to Chain B.
+
+   Example: Fully Connected Topology
+
+   In a fully connected topology, every chain has clients referencing all other chains.
+   For chains A, B, and C, the clients are created as follows:
+
+   * Chain A: Clients referencing chains B and C
+   * Chain B: Clients referencing chains A and C
+   * Chain C: Clients referencing chains A and B
+
+   This setup ensures that every chain is directly connected to every other chain, forming
+   a complete graph.
+
+   By defining the appropriate set of clients, you can establish the desired topology for
+   your tests, whether it be linear, fully connected, or any other configuration.
+*/
+
+use eyre::eyre;
+use std::collections::HashMap;
+use std::str::FromStr;
+
+use ibc_relayer::chain::handle::ChainHandle;
+use ibc_relayer::foreign_client::ForeignClient;
+
+use crate::bootstrap::binary::chain::bootstrap_foreign_client;
+use crate::error::Error;
+
+pub enum TopologyTypes {
+    Linear,
+    Full,
+}
+
+impl FromStr for TopologyTypes {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "linear" => Ok(Self::Linear),
+            "full" => Ok(Self::Full),
+            _ => Err(Error::generic(eyre!("The topology `{s}` does not exist"))),
+        }
+    }
+}
+
+pub trait Topology<Handle: ChainHandle> {
+    fn get_topology(
+        &self,
+        chain_handles: &Vec<Handle>,
+    ) -> Result<HashMap<usize, HashMap<usize, ForeignClient<Handle, Handle>>>, Error>;
+}
+
+pub struct FullyConnectedTopology;
+
+impl<Handle: ChainHandle> Topology<Handle> for FullyConnectedTopology {
+    fn get_topology(
+        &self,
+        chain_handles: &Vec<Handle>,
+    ) -> Result<HashMap<usize, HashMap<usize, ForeignClient<Handle, Handle>>>, Error> {
```

Comment:
Let's either introduce a wrapper type for this with helper methods or at least a type alias?

Reply:
Perhaps a generic `Matrix` type or similarly named?

```rust
pub struct Matrix<T> {
  matrix: HashMap<usize, HashMap<usize, T>
}

impl<T> Matrix<T> {
  // to replace calls like `let outer = map.get(x); let inner = outer.get(y)`
  pub fn get(&self, coords: (usize, usize)) -> Option<&T> {
    // ...
  }

  // to replace nested for loops
  pub fn iter(&self) -> impl Iterator<Item = &T> {
    
  }
}
```

Reply:
Good point, but I don't know if naming it a `Matrix` might be confusing. For example the linear topology for 3 chains will throw errors when accessing some values, e.g. `[0][2]`, which I wouldn't expect to be an issue in a matrix.
What about something more like `2dHashMap`?

Reply:
Yes sounds better!

Reply:
2dHashMap is not a valid identifier in Rust, so perhaps `HashMap2d` or `TwoDimHashMap`?

Reply:
Oh good catch, I have a small preference for `TwoDimHashMap` as it is slightly more explicit